include/aspect/introspection.h

```diff
@@ -29,10 +29,265 @@
+               const unsigned int multiplicity,
+               const unsigned int n_blocks)
+        : name(name), fe(fe), multiplicity(multiplicity), n_blocks(n_blocks),
+          first_component_index(-1), block_index(-1), base_index(-1)
+      {
+        Assert(n_blocks == 0
+               || n_blocks == 1
+               || n_blocks == n_components(),
+               ExcMessage("A Variable can only have 0, 1, or n_component number of blocks"));
+      }
+
+      /**
+       * Return the total number of components of this variable.
+       */
+      unsigned int n_components() const
+      {
+        return fe->n_components() * multiplicity;
+      }
+
+      /**
+       * Return a scalar FeValuesExtractor if this variable is scalar.
+       */
+      const dealii::FEValuesExtractors::Scalar &extractor_scalar() const
+      {
+        Assert(n_components()==1, ExcMessage("not a scalar FE"));
+        return extractor_scalar_;
+      }
+
+      /**
+       * Return a vector FeValuesExtractor if this variable is a vector.
+       */
+      const dealii::FEValuesExtractors::Vector &extractor_vector() const
+      {
+        Assert(n_components()==dim, ExcMessage("not a vector FE"));
+        return extractor_vector_;
+      }
+
+      std::string name;
+      std_cxx11::shared_ptr<FiniteElement<dim> > fe;
+      unsigned int multiplicity;
+      unsigned int n_blocks;
+
+      /**
+       * Initialize this variable as part of a larger FE_System wit the given
+       * indices for @p component, @p block, and @base. This function will be
+       * automatically called by IntrospectionBase and will initialize members
+       * like @p first_component_index, the extractors, and the @p component_mask.
+       */
+      void initialize(const unsigned int component_index,
+                      const unsigned int block_index_,
+                      const unsigned int base_index_)
+      {
+        block_index = block_index_;
+        first_component_index = component_index;
+        base_index = base_index_;
+        if (n_components()==1)
+          extractor_scalar_.component = component_index;
+        else
+          extractor_scalar_.component = -1;
+
+        if (n_components()==dim)
+          extractor_vector_.first_vector_component = component_index;
+        else
+          extractor_vector_.first_vector_component = -1;
+      }
+
+      /**
+       * @name Derived quantities
+       * @{
+       */
+      unsigned int first_component_index;
+      unsigned int block_index;
+      unsigned int base_index;
+
+      ComponentMask component_mask;
+      /**
+       * @}
+       */
+    private:
+      dealii::FEValuesExtractors::Scalar extractor_scalar_;
+      dealii::FEValuesExtractors::Vector extractor_vector_;
+  };
+
+  /**
+   * Class that represents a dealii::FE_System, which is created by a
+   * sorted collection of finite element variables (of type @p Variable).
+   */
+  template <int dim>
+  class IntrospectionBase
```

Comment:
Yes, I agree that the naming needs work. It represents a collection of FE variables (basically an ``FESystem``), but it in addition information about blocks (and names, etc.).

We could just call this class ``Introspection`` (in a different namespace), because the ``aspect::Introspection`` basically just adds shortcuts for the ``variable(string)`` access.